### PR TITLE
FIDEFE-4623 - Add light-dark transformer for generating web CSS

### PR DESCRIPTION
**Pull request by TGiles**
While the end result is a transformer that converts this modified JSON structure to `light-dark()` calls, there are also commits in this branch that show us how to use the parser as well.

I don't think we want to use the parser through since that converts the design-tokens JSON source for all generated platforms.
Instead it seems like we need to use transformers and formatters to change the generated file contents to fit each specific platform (if needed).

### CODE REVIEW

**Comment by TGiles** [comment]
I should probably split out the `add scss and android platforms` into its own commit.

File: toolkit/themes/shared/design-system/config.json [hunk 1]
```diff
@@ -1,12 +1,35 @@
 {
-  "source": ["design-tokens.json"],
+  "source": ["test-tokens.json"],
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.json]: Instead of having a `test-tokens.json` file and referencing it here, let's modify `design-tokens.json` to use the format you need as part of this patch

File: toolkit/themes/shared/design-system/design-tokens.json [hunk 2]
```diff
@@ -155,20 +155,18 @@
           "value": "{color.gray.05}"
         },
         "deemphasized": {
-          "light": {
-            "value": "color-mix(in srgb, currentColor 69%, transparent)"
-          },
-          "dark": {
-            "value": "color-mix(in srgb, currentColor 75%, transparent)"
-          }
+          "value": "light-dark(color-mix(in srgb, currentColor 69%, transparent),color-mix(in srgb, currentColor 75%, transparent))"
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/design-tokens.json]: I'm guessing this and the other `light-dark` reference are leftover from some experimenting? Again let's just modify this JSON so that it's in the format we need it to be in for the transform to work

File: toolkit/themes/shared/design-system/build.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+const StyleDictionary = require('style-dictionary');
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/build.js]: This might be a personal preference, but I kinda don't like how the node.js API makes you define and maintain separate `build.js` and `config.json` files. It's actually possible to just define one `config.js` file which seems to work just as well and consolidates all the configuration in one place (example [here](https://amzn.github.io/style-dictionary/#/config?id=configuration-file-formats)). What do you think of switching to that? The whole file would look something like this:

```js
// config.js
const StyleDictionary = require('style-dictionary');

StyleDictionary.registerTransform({
  // your transform here
});

StyleDictionary.registerTransformGroup({
  // your transform group here
});

module.exports = StyleDictionary.extend({
  // our config object from config.json goes here
});
```
and that should Just Work ™️ without having to fuss with `buildAllPlatforms` + having separate files

File: toolkit/themes/shared/design-system/config.json [hunk 4]
```diff
@@ -1,12 +1,35 @@
...
+          "outputReferences": true
+        }
+      }]
+    },
+    "android": {
+      "transforms": ["attribute/cti", "name/cti/snake", "color/hex", "size/remToSp", "size/remToDp"],
+      "buildPath": "build/android/src/",
+      "files": [{
+        "destination": "style_dictionary_colors.xml",
+        "format": "android/colors"
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.json]: Let's remove these since they are out of scope for our work

**Comment by hannajones** [comment]
> I should probably split out the `add scss and android platforms` into its own commit.

Whoops just saw this after leaving some comments haha 😅 

I guess we could have a separate commit for this, but its out of scope to build for these platforms for now - I don't think we're going to want to land any related code in central for them, just for CSS. I guess we could keep them around for testing purposes for now? But inevitably we're going to break the build for Android, iOS, etc. because we're going to hit a point where we would need to start adding new formats and/or transforms to get them to work.

**Review by hannajones — CHANGES_REQUESTED** [review]
Looking good! Just needs some cleanup. Lmk what you think of consolidating config into a single file

File: toolkit/themes/shared/design-system/build.js (hunk 3 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/build.js]: I don't have a strong opinion either way, let's switch to `config.js` and see how it works out for us

File: toolkit/themes/shared/design-system/design-tokens.json [hunk 5]
```diff
@@ -135,40 +101,42 @@
...
+    "brand": {
+      "value": "{color.gray.100}",
+      "darkValue": "{color.gray.05}"
+    },
+    "platform": {
+      "value": "currentColor"
+    },
     "deemphasized": {
-      "value": "color-mix(in srgb, currentColor 60%, transparent)"
+      "value": "color-mix(in srgb, currentColor 60%, transparent)",
```
Comment (TGiles) [inline toolkit/themes/shared/design-system/design-tokens.json:138]: We're going to need a way to write this token without using `color-mix` as that's probably not a thing on mobile platforms. Not something that needs to be solved right now, but good to think about.

File: toolkit/themes/shared/design-system/design-tokens.json [hunk 6]
```diff
@@ -135,40 +101,42 @@
...
+        "value": "{color.yellow.05}",
+        "darkValue": "{color.blue.80}"
       }
     }
   },
-  "platform": {
-    "value": "currentColor",
+  "text": {
+    "color": {
+      "value": "CanvasText"
```
Comment (TGiles) [inline toolkit/themes/shared/design-system/design-tokens.json:128]:
```suggestion
      "value": "CanvasText"
```
Not sure what this token should be, maybe I can just copy the format from #3 and this problem will resolve itself. Also not sure what to do about "brand" and "platform" tokens...

File: toolkit/themes/shared/design-system/config.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  const selector = args.options.selector ? args.options.selector : `:root`;
+  const indentation = args.options.indentation ? args.options.indentation: 2;
+  let modifiedDictionary = dictionary.allTokens.map(token => {
+    let value = JSON.stringify(token.value);
+    // the `dictionary` object now has `usesReference()` and
+    // `getReferences()` methods. `usesReference()` will return true if
+    // the value has a reference in it. `getReferences()` will return
+    // an array of references to the whole tokens so that you can access their
+    // names or any other attributes.
+    if (dictionary.usesReference(token.original.value) && token.original.darkValue) {
```
Comment (mstriemer) [inline toolkit/themes/shared/design-system/config.js]: Does this need to check that the light value uses a reference? Something like this seems to throw away the `light-dark` now

```
"brand": {
  "value": "black",
  "darkValue": "{color.gray.05}"
}
```
```suggestion
    if (token.original.darkValue) {
```

Trying this change locally means that the build will fail, which seems fine if we don't support the case as authored.

File: toolkit/themes/shared/design-system/config.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * to a light-dark() function. If we try to use the built-in CSS formatter,
+ * we will lose the "darkValue" or any other property that isn't "value".
+ */
+function lightDarkFormatter(args) {
+  let dictionary = args.dictionary;
+  let file = args.file;
+  const selector = args.options.selector ? args.options.selector : `:root`;
+  const indentation = args.options.indentation ? args.options.indentation: 2;
+  let modifiedDictionary = dictionary.allTokens.map(token => {
+    let value = JSON.stringify(token.value);
```
Comment (mstriemer) [inline toolkit/themes/shared/design-system/config.js]: Looks like `value` is always redefined before it's used

```suggestion
```

File: toolkit/themes/shared/design-system/config.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    if (dictionary.usesReference(token.original.value) && token.original.darkValue) {
+      // Note: make sure to use `token.original.value` because
+      // `token.value` is already resolved at this point.
+      const lightRefs = dictionary.getReferences(token.original.value);
+      const darkRefs = dictionary.getReferences(token.original.darkValue);
+      for (let i = 0; i < darkRefs.length; i++) {
+        let currentLight = lightRefs[i].name;
+        let currentDark = darkRefs[i].name;
+        value = `light-dark(var(--${currentLight}), var(--${currentDark}))`;
+        token.value = value;
```
Comment (mstriemer) [inline toolkit/themes/shared/design-system/config.js]: Feel free to keep `value` if this is going over the line limit or something, but seems like the var isn't needed

```suggestion
        token.value = `light-dark(var(--${currentLight}), var(--${currentDark}))`;
```

File: toolkit/themes/shared/design-system/config.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    // `getReferences()` methods. `usesReference()` will return true if
+    // the value has a reference in it. `getReferences()` will return
+    // an array of references to the whole tokens so that you can access their
+    // names or any other attributes.
+    if (dictionary.usesReference(token.original.value) && token.original.darkValue) {
+      // Note: make sure to use `token.original.value` because
+      // `token.value` is already resolved at this point.
+      const lightRefs = dictionary.getReferences(token.original.value);
+      const darkRefs = dictionary.getReferences(token.original.darkValue);
+      for (let i = 0; i < darkRefs.length; i++) {
```
Comment (mstriemer) [inline toolkit/themes/shared/design-system/config.js]: It seems like we'll always have 1 item in this array, if we don't then this is just hiding a bug I think

This got me wondering *why* it's even an array though. Looks like we could do something like:

```json
"brand": {
  "value": "",
  "lightDarkValue": ["{color.gray.100}", "{color.gray.05}"]
}
```

Which is now making me wonder if we could transform directly on the `value` key?

```json
{
  "text": {
    "color": {
      "value": "CanvasText"
    },
    "brand": {
      "value": ["{color.gray.100}", "{color.gray.05}"]
    }
  }
}
```

File: toolkit/themes/shared/design-system/config.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        let currentDark = darkRefs[i].name;
+        value = `light-dark(var(--${currentLight}), var(--${currentDark}))`;
+        token.value = value;
+      }
+    }
+    return token;
+  });
+  dictionary.allTokens = modifiedDictionary;
+  let mappedValues = modifiedDictionary.map(element => {
+    let joinedValue = `${" ".repeat(indentation)}--${element.name}: ${element.value};`
```
Comment (mstriemer) [inline toolkit/themes/shared/design-system/config.js]: Nit: This is gonna run repeat on every iteration but doesn't look like it's gonna change

```suggestion
  let indent = " ".repeat(indentation);
  let mappedValues = modifiedDictionary.map(
    element => `${indent}--${element.name}: ${element.value};`
  );
```

File: toolkit/themes/shared/design-system/design-tokens.json (hunk 5 above)
Comment (mstriemer) [reply toolkit/themes/shared/design-system/design-tokens.json:138]: Maybe we could key this using the array values?

```json
"deemphasized": {
  "value": ["color-mix", "in srgb", "currentColor 60%", "transparent"]
}
```

I believe that would let us try to come up with a solution in a transformer for mobile, and we could potentially even have a generic transformer for light-dark, etc

```json
"critical": {
  "value": ["light-dark", "{color.red.05}", "{color.red.80}"]
}
```

**Review by mstriemer — CHANGES_REQUESTED** [review]
Looking good, could definitely discuss some of these more or plan for follow ups 👍

File: toolkit/themes/shared/design-system/build.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+// APPLY THE CONFIGURATION
+// IMPORTANT: the registration of custom transforms
+// needs to be done _before_ applying the configuration
+const StyleDictionaryExtended = StyleDictionary.extend(__dirname + '/config.js');
+
+// FINALLY, BUILD ALL THE PLATFORMS
+StyleDictionaryExtended.buildAllPlatforms();
+
+console.log('\n==============================================');
+console.log('\nBuild completed!');
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/build.js]: now that things are in `config.js` we don't need this file at all anymore - you can just delete it

File: toolkit/themes/shared/design-system/config.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    return joinedValue;
+  }).join("\n");
+  return `${fileHeader({ file })}\n${selector} {\n${mappedValues}\n}`;
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
return StyleDictionary.format["css/variables"]({ ...args, dictionary })
```

You could actually replace all of this by just using the built in formatter, in which case you also don't need to worry about reading the `selector` or `indentation` args (seems ok for now since we're not modifying them).

File: toolkit/themes/shared/design-system/config.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  format: {
+    "css/variables/light-dark": lightDarkFormatter,
+  },
+  platforms: {
+    css: {
+      transformGroup: "css",
+      buildPath: "build/css/",
+      files: [{
+        destination: "light-dark.css",
+        format: "css/variables/light-dark",
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
        format: "css/variables/light-dark",
        options: {
          outputReferences: true,
        }
```

We'll still need this for cases where we have variables that aren't referenced in `light-dark`

File: toolkit/themes/shared/design-system/design-tokens.json [hunk 15]
```diff
@@ -135,40 +101,42 @@
...
-            "value": "color-mix(in srgb, currentColor 75%, transparent)"
-          }
-        }
+    "white": {
+      "value": "#ffffff"
+    },
+    "background": {
+      "critical": {
+        "value": "{color.red.05}",
+        "darkValue": "{color.red.80}"
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/design-tokens.json]:
```suggestion
        "dark": "{color.red.80}"
```

nit: I thought we were gonna drop the `Value` suffix to make things more concise? I still don't have strong opinions about this but we should probably pick one and stick with it (happy to update my patch if we want to use it)

File: toolkit/themes/shared/design-system/design-tokens.json (hunk 6 above)
Comment (hannajones) [reply toolkit/themes/shared/design-system/design-tokens.json:128]: I think we're just gonna stick with using an empty string + `default` to get around some of the weirdness for now since it makes things work, then re-visit those choices later.

File: toolkit/themes/shared/design-system/config.js (hunk 10 above)
Comment (hannajones) [reply toolkit/themes/shared/design-system/config.js]: I do wonder if it'll make things easier for iOS / Android to have the `dark` values spelled out explicitly since they break things up into themes...though I guess that's still many steps down the line and having to grab the second value in an array isn't complicated 🤷🏻

**Review by hannajones — CHANGES_REQUESTED** [review]
Working well, just a few more nits/pieces of cleanup!

I'm curious what motivated the switch from a transform to a formatter? It seems like the transforms [get applied before everything gets formatted](https://amzn.github.io/style-dictionary/#/architecture), so it could be a nice way of breaking things up.

On the other hand it shouldn't be a big deal to integrate this with the other formatting logic, I'm just kinda curious if the transform didn't work for some reason.

File: toolkit/themes/shared/design-system/build.js (hunk 12 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/build.js]: @hannajones but then how do we run the build? I tried `node config.js` and nothing would happen lol. What am I missing?

Comment (hannajones) [reply toolkit/themes/shared/design-system/build.js]: I just run `style-dictionary build` from the command line and don't bother with node 🤷🏻 If you want you can add something like this to `package.json` to make it work with `npm run build`:

```json
"scripts": {
  "build": "style-dictionary build"
}
```

File: toolkit/themes/shared/design-system/config.js (hunk 10 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/config.js]: I think we can use the current W3C design tokens structure and use the composite form (https://design-tokens.github.io/community-group/format/#composite-design-token) to handle this case. For example:
```json
{
"foo": {
    "value": {
      "light": "white",
      "dark": "black",
      "forcedColors": "CanvasText",
    }
  },
  "bar": {
    "value": {
      "light": "{color.gray.100}",
      "dark": "{color.gray.05}"
    }
  }
}
```
and then use a formatter to convert the values accordingly.

File: toolkit/themes/shared/design-system/config.js (hunk 13 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/config.js]: Oh this is great, thanks for that assist! I was really hoping I didn't need to do this "custom" formatting 😅

**Comment by TGiles** [comment]
> I'm curious what motivated the switch from a transform to a formatter? It seems like the transforms [get applied before everything gets formatted](https://amzn.github.io/style-dictionary/#/architecture), so it could be a nice way of breaking things up.

@hannajones  So if we try to do the light-dark transforms, we'll lose the variable references once the default CSS formatter does its magic (which is why I had the fileHeader, mappedValues block). IIRC, the `outputReferences` would drop the entire transformed token which isn't too useful for us lol.

File: toolkit/themes/shared/design-system/build.js (hunk 12 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/build.js]: oh duh, I should have been using `style-dictionary` to build it; totally forgot about it for some reason. Thanks!

**Comment by TGiles** [comment]
renamed the branch and the PR closed, let me fix that.

File: toolkit/themes/shared/design-system/config.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/* eslint-env node */
+
+const StyleDictionary = require("style-dictionary");
+const { fileHeader } = StyleDictionary.formatHelpers;
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
```

Looks like we're not using this anymore so we can remove it (though I have a feeling we'll need to use it again down the line to add the MPL - a problem for another day 😅)

File: toolkit/themes/shared/design-system/config.js [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * CSS formatter that converts tokens that have a light and dark value
+ * to a light-dark() function. If we try to use the built-in CSS formatter,
+ * we will lose the "darkValue" or any other property that isn't "value".
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
 * we will lose the "dark" value or any other property that isn't "value".
```
nit: this is the nittiest of nit picks just to match what we have in the JSON now

File: toolkit/themes/shared/design-system/config.js [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+    // `getReferences()` methods. `usesReference()` will return true if
+    // the value has a reference in it. `getReferences()` will return
+    // an array of references to the whole tokens so that you can access their
+    // names or any other attributes.
+    if (token.original.dark) {
+      // Note: make sure to use `token.original.value` because
+      // `token.value` is already resolved at this point.
+      const lightRefs = dictionary.getReferences(token.original.value);
+      const darkRefs = dictionary.getReferences(token.original.dark);
+      if (lightRefs.length > 0 || darkRefs.length > 0) {
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
      if (lightRefs.length || darkRefs.length) {
```
ESLint doesn't seem to like the `> 0` check

File: toolkit/themes/shared/design-system/config.js [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+          let currentDarkName = darkRefs[i]?.name;
+          let lightValue = currentLightName ? 
+          `var(--${currentLightName})` : `${token.original.value}`;
+          let darkValue = currentDarkName ?
+          `var(--${currentDarkName})` : `${token.original.dark}`;
+          token.value = `light-dark(${lightValue}, ${darkValue})`
+        }
+      } else {
+        value = `light-dark(${token.original.value}, ${token.original.dark})`;
+        token.value = value;
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
        token.value = `light-dark(${token.original.value}, ${token.original.dark})`;
```
Looks like `value` is no longer defined above so we either need to do `let value =`, or just make this a one liner

**Review by hannajones — CHANGES_REQUESTED** [review]
This is no longer working for me locally and I think I may have led you astray by suggesting this for formatting:

```js
StyleDictionary.format["css/variables"]({...args, dictionary});
```

It seems that when combined with the `outputReferences` option it ends up overwriting the `light-dark` values for some reason ☹️ You might have to go back to the manual formatting, or you might be able to use the `formattedVariables` [helper](https://amzn.github.io/style-dictionary/#/formats?id=formattedvariables) in some way

File: toolkit/themes/shared/design-system/config.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+      if (lightRefs.length > 0 || darkRefs.length > 0) {
+        for (let i = 0; i < darkRefs.length; i++) {
+          let currentLightName = lightRefs[i]?.name;
+          let currentDarkName = darkRefs[i]?.name;
+          let lightValue = currentLightName ? 
+          `var(--${currentLightName})` : `${token.original.value}`;
+          let darkValue = currentDarkName ?
+          `var(--${currentDarkName})` : `${token.original.dark}`;
+          token.value = `light-dark(${lightValue}, ${darkValue})`
+        }
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
          let currentLightName = lightRefs[0]?.name;
          let currentDarkName = darkRefs[0]?.name;
          let lightValue = currentLightName ? 
          `var(--${currentLightName})` : `${token.original.value}`;
          let darkValue = currentDarkName ?
          `var(--${currentDarkName})` : `${token.original.dark}`;
          token.value = `light-dark(${lightValue}, ${darkValue})`
```

Looking at [the source](https://github.com/amzn/style-dictionary/blob/17f4cb2f30bd002dfd55d6ef8c5bee4138de8d64/lib/utils/references/getReferences.js#L31) it seems like `getReferences` only returns an array for cases where a single value interpolates multiple variables, something like:

```json
"light": "light-dark({color-red-05}, {color-red-80})"
```
since that's not a thing for us at the moment we can probably avoid looping (plus using the length of `darkRefs` could hypothetically cause a bug if we had multiple interpolated variables in the light value but not the dark value)

File: toolkit/themes/shared/design-system/config.js (hunk 16 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/config.js]: Unfortunately I still need the file header since I have to do the formatting myself instead of being able to use "css/variables" from style dictionary.

File: toolkit/themes/shared/design-system/config.js (hunk 20 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/config.js]: Works for me, thanks for looking into the source!

File: toolkit/themes/shared/design-system/config.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+      // `token.value` is already resolved at this point.
+      const lightRefs = dictionary.getReferences(token.original.value);
+      const darkRefs = dictionary.getReferences(token.original.dark);
+      let currentLightName = lightRefs[0]?.name;
+      let currentDarkName = darkRefs[0]?.name;
+      let lightValue = currentLightName ?
+        `var(--${currentLightName})` : `${token.original.value}`;
+      let darkValue = currentDarkName ?
+        `var(--${currentDarkName})` : `${token.original.dark}`;
+      token.value = `light-dark(${lightValue}, ${darkValue})`
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
      token.value = token.originalValue =`light-dark(${lightValue}, ${darkValue})`
```

So uh...I think this might be the key to solving all of our problems...

Testing locally and overwriting the "original" value means we can use `outputReferences` without worrying about `light-dark` getting overwritten 🎉

**Review by hannajones — COMMENTED** [review]
@TGiles I'm so sorry for all the back and forth on this...I'm learning as I go here lol

We both observed that `outputReferences` messes with our attempts to overwrite `value` in some cases. Messing around with it some more it seems like if we overwrite the original value all our problems are solved. Happy to pair on this tomorrow if you want to make sure we're on the same page, just lmk 🙂

**Comment by TGiles** [comment]
@hannajones @mstriemer @jsimplicio I think we're almost there for this PR, very happy with the progress we're making with this!

File: toolkit/themes/shared/design-system/config.js (hunk 21 above)
Comment (TGiles) [reply toolkit/themes/shared/design-system/config.js]: This is totally the key, thank you for figuring that out!

File: toolkit/themes/shared/design-system/config.js [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+        // modify the original value and everything works like magic
+        token.original.value = lightDarkValue;
+        return lightDarkValue;
+      },
+    },
+  },
+  platforms: {
+    css: {
+      transforms: [
+        ...StyleDictionary.transformGroup["css"],
```
Comment (hannajones) [inline toolkit/themes/shared/design-system/config.js]:
```suggestion
        ...StyleDictionary.transformGroup.css,
```
ESLint seems to want dot notation here 🤷🏻

**Review by hannajones — APPROVED** [review]
LGTM, thanks for sticking with this through all the back and forth 🤠

**Review by mstriemer — APPROVED** [review]
This looks great! Awesome to see all the experimenting and iterating and it coming out nice and tidy like this 👍
